pkg/clioptions/clusterinfo/wait_for_stable_cluster.go

```diff
@@ -0,0 +1,183 @@
+			{
+				Name: testName,
+			},
+		}, nil
+	}
+
+	configClient, err := configclient.NewForConfig(config)
+	if err != nil {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: fmt.Sprintf("error creating config client: %v", err),
+				},
+			},
+		}, err
+	}
+
+	interval := 10 * time.Second
+	timeout := 10 * time.Minute
+	minimumStablePeriod := 3 * time.Minute
+	recoveredOperators := map[string]string{}
+	unstableOperators := map[string]string{}
+	var stabilityStarted *time.Time
+	pollingStarted := time.Now()
+	waitErr := wait.PollUntilContextTimeout(ctx, interval, timeout, true, func(waitCtx context.Context) (bool, error) {
+		operators, err := configClient.ConfigV1().ClusterOperators().List(waitCtx, metav1.ListOptions{})
+		if err != nil {
+			if waitCtx.Err() == context.DeadlineExceeded {
+				return false, fmt.Errorf("some operators were unstable for too long")
+			}
+			stabilityStarted = nil
+			return false, fmt.Errorf("failed to list clusteroperators: %v", err)
+		}
+		now := time.Now()
+		for _, operator := range operators.Items {
+			if unstableReason := unstableOperatorReason(&operator); len(unstableReason) > 0 {
+				if _, ok := unstableOperators[operator.Name]; !ok {
+					unstableOperators[operator.Name] = unstableReason
+				}
+			} else {
+				if unstableReason, ok := unstableOperators[operator.Name]; ok {
+					recoveredOperators[operator.Name] = unstableReason
+					delete(unstableOperators, operator.Name)
+				}
+			}
+		}
+		if len(unstableOperators) > 0 {
+			stabilityStarted = nil
+			return false, nil
+		}
+		if stabilityStarted == nil {
+			stabilityStarted = &now
+		}
+		timeStable := now.Sub(*stabilityStarted)
+		if timeStable < minimumStablePeriod {
+			return false, nil
+		}
+		return true, nil
+	})
+	timingMsg := ""
+	if stabilityStarted != nil {
+		timingMsg = fmt.Sprintf(", unstable duration %s", stabilityStarted.Sub(pollingStarted))
+	}
+	if waitErr != nil {
+		msg := fmt.Sprintf("error waiting for cluster operators to become stable: %v", waitErr)
+		if len(unstableOperators) > 0 {
+			msg += fmt.Sprintf("\nunstable operators that never recovered\n")
+			msg += summerizeUnstableOperators(unstableOperators)
+		}
+		if len(recoveredOperators) > 0 {
+			msg += fmt.Sprintf("\nunstable operators that recovered%s\n", timingMsg)
+			msg += summerizeUnstableOperators(recoveredOperators)
+		}
+		// Flake for now.
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+			},
+			{
+				Name: testName,
+				FailureOutput: &junitapi.FailureOutput{
+					Output: msg,
+				},
+			},
+		}, waitErr
+	}
+	if len(recoveredOperators) == 0 {
+		return []*junitapi.JUnitTestCase{
+			{
+				Name: testName,
+			},
+		}, nil
+	}
+	// Some operators recovered from unstable conditions before timeout
+	msg := fmt.Sprintf("some operators were unstable but recovered%s\n%s", timingMsg, summerizeUnstableOperators(recoveredOperators))
```

Comment:
When we discussed last I was thinking we would have two separate tests

1.  "Cluster should be stable before test is started"
2.  "Cluster should become stable before upgrade is started" - or similar

They would both flake instead of fail, it looks like this has one test that flakes for different reasons.  I believe we wanted to be able to see how often the cluster needed additional time to stabilize compared to how often does it not stabilize at all.